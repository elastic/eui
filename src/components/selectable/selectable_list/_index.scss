@import 'selectable_list';
@import 'selectable_list_item';
