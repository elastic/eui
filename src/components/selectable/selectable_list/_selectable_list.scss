// Expand height of list via flex box
.euiSelectableList-fullHeight {
  flex-grow: 1;
}

.euiSelectableList-bordered {
  overflow: hidden;
  border: $euiBorderThin;
  border-radius: $euiBorderRadius;
}

.euiSelectableList__list {
  @include euiOverflowShadow;
  @include euiScrollBar;
}

.euiSelectableList__groupLabel {
  @include euiTitle('xxxs');
  display: flex;
  align-items: center;
  border-bottom: $euiSelectableListItemBorder;
}
