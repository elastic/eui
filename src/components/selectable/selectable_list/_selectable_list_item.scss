.euiSelectableListItem {
  @include euiFontSizeS;

  padding: $euiSizeXS $euiSizeM;
  display: block; // Necessary to make sure it doesn't force the whole popover to be too wide
  width: 100%;
  text-align: left;
  color: $euiTextColor;

  &:not(:last-of-type){
    border-bottom: 1px solid darken($euiColorLightestShade, 2%);
  }

  &:hover,
  &:focus {
    text-decoration: underline;
  }

  &:focus,
  &-isFocused {
    cursor: pointer;
    color: $euiColorPrimary;
    background-color: $euiFocusBackgroundColor;
  }

  &[disabled] {
    color: $euiColorMediumShade;
    cursor: not-allowed;

    &:hover {
      text-decoration: none;
    }
  }
}

.euiSelectableListItem__content {
  display: flex;
  align-items: center;
}

.euiSelectableListItem__icon {
  margin-right: $euiSizeXS;
}

.euiSelectableListItem__text {
  @include euiTextTruncate;
}
