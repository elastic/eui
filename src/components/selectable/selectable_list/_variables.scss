$euiSelectableListItemBorder: 1px solid darken($euiColorLightestShade, 2%);
$euiSelectableListItemPadding: $euiSizeXS $euiSizeM;
