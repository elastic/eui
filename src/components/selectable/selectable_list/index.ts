export { EuiSelectableList } from './selectable_list';
export { EuiSelectableListItem } from './selectable_list_item';
