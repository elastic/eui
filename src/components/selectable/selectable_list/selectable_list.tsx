import React, { Component, HTMLAttributes, ReactNode } from 'react';
import classNames from 'classnames';
import { CommonProps } from '../../common';
// eslint-disable-next-line import/named
import { List, AutoSizer, ListProps } from 'react-virtualized';
import { htmlIdGenerator } from '../../../services';
import { EuiSelectableListItem } from './selectable_list_item';
// @ts-ignore
import { EuiHighlight } from '../../highlight';
import { Option } from '../types';

export type EuiSelectableSingleOptionProps = 'always' | boolean;

// Consumer Configurable Props via `EuiSelectable.listProps`
export type EuiSelectableOptionsListProps = HTMLAttributes<HTMLDivElement> &
  CommonProps & {
    /**
     * The index of the option to be highlighted as pseudo-focused;
     * Good for use when only one selection is allowed and needing to open
     * directly to that option
     */
    activeOptionIndex?: number;
    /**
     *  The height of each option in pixels. Defaults to `32`
     */
    rowHeight: number;
    /**
     * Show the check/cross selection indicator icons
     */
    showIcons?: boolean;
    singleSelection?: EuiSelectableSingleOptionProps;
    /**
     * Any props to send specifically to the react-virtualized `List`
     */
    virtualizedProps?: ListProps;
    /**
     * Adds a border around the list to indicate the bounds;
     * Useful when the list scrolls, otherwise use your own container
     */
    bordered?: boolean;
  };

export type EuiSelectableListProps = EuiSelectableOptionsListProps & {
  /**
   * All possible options
   */
  options: Option[];
  /**
   * Filtered options list (if applicable)
   */
  visibleOptions?: Option[];
  /**
   * Search value to highlight on the option render
   */
  searchValue: string;
  /**
   * Returns the array of options with altered checked state
   */
  onOptionClick: (options: Option[]) => void;
  /**
   * Custom render for the label portion of the option;
   * Takes (option, searchValue), returns ReactNode
   */
  renderOption?: (option: Option, searchValue: string) => ReactNode;
  /**
   * Sets the max height in pixels or pass `full` to allow
   * the whole group to fill the height of its container and
   * allows the list grow as well
   */
  height?: number | 'full';
  /**
   * Allow cycling through the on, off and undefined state of option.checked
   * and not just on and undefined
   */
  allowExclusions?: boolean;
  rootId?: (appendix?: string) => string;
};

export class EuiSelectableList extends Component<EuiSelectableListProps> {
  static defaultProps = {
    rowHeight: 32,
    searchValue: '',
  };

  rootId = this.props.rootId || htmlIdGenerator();

  constructor(props: EuiSelectableListProps) {
    super(props);
  }

  render() {
    const {
      className,
      options,
      searchValue,
      onOptionClick,
      renderOption,
      height: forcedHeight,
      virtualizedProps,
      rowHeight,
      activeOptionIndex,
      rootId,
      showIcons,
      singleSelection,
      visibleOptions,
      allowExclusions,
      bordered,
      ...rest
    } = this.props;

    const optionArray = visibleOptions || options;

    const heightIsFull = forcedHeight === 'full';

    let calculatedHeight = (heightIsFull ? false : forcedHeight) as
      | false
      | number
      | undefined;

    // If calculatedHeight is still undefined, then calculate it
    if (calculatedHeight === undefined) {
      const maxVisibleOptions = 7;
      const numVisibleOptions = optionArray.length;
      const numVisibleMoreThanMax = optionArray.length > maxVisibleOptions;

      if (numVisibleMoreThanMax) {
        // Show only half of the last one to indicate there's more to scroll to
        calculatedHeight = (maxVisibleOptions - 0.5) * rowHeight;
      } else {
        calculatedHeight = numVisibleOptions * rowHeight;
      }
    }

    const classes = classNames(
      'euiSelectableList',
      {
        'euiSelectableList-fullHeight': heightIsFull,
        'euiSelectableList-bordered': bordered,
      },
      className
    );

    return (
      <div className={classes} {...rest}>
        <AutoSizer disableHeight={!heightIsFull}>
          {({ width, height }) => (
            <List
              id={this.rootId('listbox')}
              className="euiSelectableList__list"
              role="listbox"
              width={width}
              height={calculatedHeight || height}
              rowCount={optionArray.length}
              rowHeight={rowHeight}
              scrollToIndex={activeOptionIndex}
              {...virtualizedProps}
              rowRenderer={({ key, index, style }) => {
                const option = optionArray[index];
                const {
                  label,
                  isGroupLabel,
                  checked,
                  disabled,
                  prepend,
                  append,
                  ref,
                  ...optionRest
                } = option;
                if (isGroupLabel) {
                  return (
                    <div
                      className="euiSelectableList__groupLabel"
                      key={key}
                      style={style}
                      {...optionRest}>
                      {prepend}
                      {label}
                      {append}
                    </div>
                  );
                }
                return (
                  <EuiSelectableListItem
                    id={this.rootId(`_option-${index}`)}
                    style={style}
                    key={option.label.toLowerCase()}
                    onClick={() => this.onAddOrRemoveOption(option)}
                    ref={ref ? ref.bind(null, index) : undefined}
                    isFocused={activeOptionIndex === index}
                    title={label}
                    showIcons={showIcons}
                    checked={checked}
                    disabled={disabled}
                    prepend={prepend}
                    append={append}
                    {...optionRest}>
                    {renderOption ? (
                      renderOption(option, searchValue)
                    ) : (
                      <EuiHighlight search={searchValue}>{label}</EuiHighlight>
                    )}
                  </EuiSelectableListItem>
                );
              }}
            />
          )}
        </AutoSizer>
      </div>
    );
  }

  onAddOrRemoveOption = (option: Option) => {
    if (option.disabled) {
      return;
    }

    const { allowExclusions } = this.props;

    if (option.checked === 'on' && allowExclusions) {
      this.onExcludeOption(option);
    } else if (option.checked === 'on' || option.checked === 'off') {
      this.onRemoveOption(option);
    } else {
      this.onAddOption(option);
    }
  };

  private onAddOption = (addedOption: Option) => {
    const { onOptionClick, options, singleSelection } = this.props;

    const updatedOptions = options.map(option => {
      // if singleSelection is enabled, uncheck any selected option(s)
      const updatedOption = { ...option };
      if (singleSelection) {
        delete updatedOption.checked;
      }

      // if this is the now-selected option, check it
      if (option === addedOption) {
        updatedOption.checked = 'on';
      }

      return updatedOption;
    });

    onOptionClick(updatedOptions);
  };

  private onRemoveOption = (removedOption: Option) => {
    const { onOptionClick, singleSelection, options } = this.props;

    const updatedOptions = options.map(option => {
      const updatedOption = { ...option };

      if (option === removedOption && singleSelection !== 'always') {
        delete updatedOption.checked;
      }

      return updatedOption;
    });

    onOptionClick(updatedOptions);
  };

  private onExcludeOption = (excludedOption: Option) => {
    const { onOptionClick, options } = this.props;
    excludedOption.checked = 'off';

    const updatedOptions = options.map(option => {
      const updatedOption = { ...option };

      if (option === excludedOption) {
        updatedOption.checked = 'off';
      }

      return updatedOption;
    });

    onOptionClick(updatedOptions);
  };
}
