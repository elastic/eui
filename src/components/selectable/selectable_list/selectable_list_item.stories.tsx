/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';

import { OPTION_CHECKED_STATES } from '../selectable_option';
import {
  EuiSelectableListItem,
  EuiSelectableListItemProps,
} from './selectable_list_item';

const meta: Meta<EuiSelectableListItemProps> = {
  title: 'Forms/EuiSelectable/EuiSelectableList/EuiSelectableListItem',
  component: EuiSelectableListItem,
  argTypes: {
    checked: {
      control: 'radio',
      options: [undefined, ...OPTION_CHECKED_STATES],
    },
    append: {
      control: 'boolean',
      description: 'Use the control to toggle showing an appended example',
      mapping: {
        true: <span>Append</span>,
        false: undefined,
      },
    },
    prepend: {
      control: 'boolean',
      description: 'Use the control to toggle showing an prepended example',
      mapping: {
        true: <span>Prepend</span>,
        false: undefined,
      },
    },
  },
  args: {
    showIcons: true,
    paddingSize: 's',
    onFocusBadge: true,
    textWrap: 'truncate',
    // set up for easier testing/QA
    allowExclusions: false,
    searchable: false,
    disabled: false,
    isFocused: false,
    append: false,
    prepend: false,
  },
};

export default meta;
type Story = StoryObj<EuiSelectableListItemProps>;

export const Playground: Story = {
  args: {
    children: 'Selectable list item',
  },
};
