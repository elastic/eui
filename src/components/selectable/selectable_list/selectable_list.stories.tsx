/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import type { Meta, StoryObj } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import {
  enableFunctionToggleControls,
  moveStorybookControlsToCategory,
} from '../../../../.storybook/utils';
import { EuiSelectableOption } from '../selectable_option';
import { EuiSelectableList, EuiSelectableListProps } from './selectable_list';

const options: EuiSelectableOption[] = [
  {
    label: 'Titan',
    'data-test-subj': 'titanOption',
  },
  {
    label: 'Enceladus is disabled',
    disabled: true,
  },
  {
    label: 'Mimas',
    checked: 'on',
  },
  {
    label: 'Dione',
  },
  {
    label: 'Iapetus',
    checked: 'on',
  },
  {
    label: 'Phoebe',
  },
  {
    label: 'Rhea',
  },
  {
    label:
      "Pandora is one of Saturn's moons, named for a Titaness of Greek mythology",
  },
  {
    label: 'Tethys',
  },
  {
    label: 'Hyperion',
  },
];

const meta: Meta<EuiSelectableListProps<{}>> = {
  title: 'Forms/EuiSelectable/EuiSelectableList/EuiSelectableList',
  component: EuiSelectableList,
  argTypes: {
    height: { control: 'number' },
    isPreFiltered: { control: 'boolean' },
    singleSelection: { control: 'boolean' },
  },
  args: {
    textWrap: 'truncate',
    paddingSize: 's',
    onFocusBadge: true,
    showIcons: true,
    // set up for easier testing/QA
    listId: '',
    allowExclusions: false,
    bordered: false,
    isPreFiltered: false,
    isVirtualized: false,
    searchable: false,
    singleSelection: true,
  },
};
enableFunctionToggleControls(meta, ['onOptionClick', 'setActiveOptionIndex']);
moveStorybookControlsToCategory(
  meta,
  [
    'allowExclusions',
    'onFocusBadge',
    'paddingSize',
    'searchable',
    'showIcons',
    'textWrap',
  ],
  'EuiSelectableListItem props'
);

export default meta;
type Story = StoryObj<EuiSelectableListProps<{}>>;

export const Playground: Story = {
  args: {
    options,
    activeOptionIndex: 0,
    makeOptionId: (index) => `selectable_list_item-${index}`,
    // ensuring that onOptionClick triggers an action as it's
    // only called through setActiveOptionIndex callback
    setActiveOptionIndex: (index, callback) => {
      callback?.();
      action('setActiveOptionIndex')(index);
    },
  },
};
