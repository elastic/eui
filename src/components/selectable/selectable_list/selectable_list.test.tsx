import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { EuiSelectableList } from './selectable_list';

describe('EuiSelectableList', () => {
  test('is rendered', () => {
    const component = render(
      <EuiSelectableList {...requiredProps} />
    );

    expect(component)
      .toMatchSnapshot();
  });
});
