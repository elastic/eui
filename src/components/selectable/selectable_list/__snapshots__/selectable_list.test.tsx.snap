// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiSelectableListItem is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props activeOptionIndex 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props allowExclusions 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props bordered 1`] = `
<div
  class="euiSelectableList euiSelectableList-bordered"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props height is forced 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:200px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props height is full 1`] = `
<div
  class="euiSelectableList euiSelectableList-fullHeight"
>
  <div
    style="overflow:visible;height:0;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:0;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:auto"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props renderOption 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props rowHeight 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:120px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props searchValue 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props searchValue 2`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props showIcons can be turned off 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props singleSelection can be forced so that at least one must be selected 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props singleSelection can be turned on 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:192px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`EuiSelectableListItem props visibleOptions 1`] = `
<div
  class="euiSelectableList"
>
  <div
    style="overflow:visible;width:0"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List euiSelectableList__list"
      id="htmlid_listbox"
      role="listbox"
      style="box-sizing:border-box;direction:ltr;height:128px;position:relative;width:0;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden"
      tabindex="0"
    />
  </div>
</div>
`;
